src/lib.rs

```diff
@@ -287,7 +286,10 @@ pub trait OptionExt<T> {
 /// # Safety
 ///
 /// It must be impossible to produce a value of this type.
-pub unsafe trait Never {}
+pub unsafe trait Never {
+    /// Coerces this type into the never type, `!`.
+    fn never(self) -> !;
```

Comment:
The primitive never type `!` is currently unstable. I'd love to use it but rancor has a hard requirement to be 100% stable. We can still keep the `match self {}`s without needing this `never` method, except in the generic context.

If we get it stabilized someday, I think the way I'd like this to work would be:

```rust
#[inline(always)]
const fn unreachable_checked<T: Never>(_: T) -> ! { unsafe { unreachable_unchecked() } }
```

That way the trait can remain empty and we can just call `unreachable_checked(e)` to get the never-ing we want, and in a const context.

In a truly perfect world, we'd want a builtin `Never` auto trait with the special property that instead of all fields of a type having to implement `Never` for the type to implement `Never`, only one `Never` field must be present at any time for it to implement `Never`.

Reply:
The never type is currently stable only in function return values. My change compiles on stable.

Reply:
Oh dang really? I didn't realize that, sorry for the quick assumption / quick merge. Would you want to add that `unreachable_checked` function in another PR? I'm happy to add it too if you're currently busy.